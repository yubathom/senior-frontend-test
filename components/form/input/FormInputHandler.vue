<template>
  <input
    class="w-full rounded border border-solid p-2 bg-white shadow-micro border-primary-darkblue transition-colors duration-300"
    :type="type"
    :class="{
      'border-accent-blue': valid && focused,
      'border-accent-red': error
    }"
    :name="name"
    :minlength="minlength"
    :placeholder="placeholder"
    v-model="content"
    @focus.once="pristine = false"
    @focus="focused = true"
    @blur="handleInput(error, _uid, content, true)"
    @keyup.prevent="handleInput(error, _uid, content, false)"
  />
</template>
<script>
export default {
  name: 'FormInputHandler',
  props: {
    name: {
      type: String,
      required: true
    },
    minlength: {
      type: Number,
      required: true
    },
    type: {
      type: String,
      required: true
    },
    placeholder: {
      type: String,
      default: ''
    },
    initialValue: {
      type: String,
      default: ''
    }
  },
  data () {
    return {
      content: '',
      focused: false,
      pristine: true
    }
  },
  computed: {
    valid () {
      return {
        text: this.content.length >= this.minlength,
        email: (() => {
          const emailPattern = /\S+@\S+\.\S+/
          return emailPattern.test(this.content)
        })(),
        phone: (() => {
          const phonePattern = /^[+]*[(]{0,1}[0-9]{1,3}[)]{0,1}[-\s\./0-9]*$/g
          return phonePattern.test(this.content)
        })()
      }[this.type]
    },
    error () {
      return this.pristine || this.valid ? false : true
    }
  },
  created () {
    this.content = this.initialValue
  },
  methods: {
    handleInput (errorStatus, uid, content, blur) {
      if (blur) this.focused = false
      this.$emit('checkInput', errorStatus, uid, content)
    }
  }
}
</script>
