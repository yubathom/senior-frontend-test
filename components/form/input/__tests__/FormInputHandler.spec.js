import { mount } from '@vue/test-utils'
import FormInputHandler from '../FormInputHandler'

describe('FormInputHandler', () => {
  const propsData = {
    name: 'test',
    minlength: 3,
    type: 'text',
    placeholder: 'foo',
    initialValue: 'bar'
  }

  const wrapper = propsData => mount(FormInputHandler, { propsData })

  it('is properly rendering', () => {
    expect(wrapper(propsData).html()).toMatchSnapshot()
  })

  it('test a valid text input', () => {
    const { vm } = wrapper(propsData)
    expect(vm.valid).toBe(true)
    expect(vm.error).toBe(false)
  })
  it('test a invalid text input', () => {
    let currentProp = propsData
    currentProp.initialValue = 'b' // smaller than current minlength -> 3

    const { vm } = wrapper(currentProp)
    expect(vm.valid).toBe(false)
  })
  it('test a valid email input', () => {
    let currentProp = propsData
    currentProp.type = 'email'
    currentProp.initialValue = 'email@email.com'

    const { vm } = wrapper(currentProp)
    expect(vm.valid).toBe(true)
  })
  it('test a invalid email input', () => {
    let currentProp = propsData
    currentProp.type = 'email'
    currentProp.initialValue = 'not a email'

    const { vm } = wrapper(currentProp)
    expect(vm.valid).toBe(false)
  })
  it('test valid phone numbers', () => {
    let currentProp = propsData
    currentProp.type = 'phone'

    const validNumbers = [
      '(333)333-333',
      '123456789',
      '333 333 333'
    ]

    validNumbers.forEach(num => {
      currentProp.initialValue = num
      const { vm } = wrapper(currentProp)
      expect(vm.valid).toBe(true)
    })
  })
  it('test invalid phone numbers', () => {
    let currentProp = propsData
    currentProp.type = 'phone'

    const invalidNumbers = [
      'qweqweasd',
      '1234qwe789',
      '333@333&*(333'
    ]

    invalidNumbers.forEach(num => {
      currentProp.initialValue = num
      const { vm } = wrapper(currentProp)
      expect(vm.valid).toBe(false)
    })
  })
})
