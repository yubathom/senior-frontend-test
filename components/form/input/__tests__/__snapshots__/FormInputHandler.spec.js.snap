// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInputHandler is properly rendering 1`] = `<input name="test" minlength="3" placeholder="foo" type="text" class="w-full rounded border border-solid p-2 bg-white shadow-micro border-primary-darkblue transition-colors duration-300">`;
